### ISSUE TITLE
Segfault when saving PNG too often

### ISSUE THREAD

**Issue by fonsp** [JuliaIO/PNGFiles.jl#32]
Here is a MWE. I tried to make it smaller, but it seems like all of these steps are needed:

```julia
import Pkg
Pkg.activate(mktempdir())
Pkg.add([
    Pkg.PackageSpec(name="Images", version="0.23.3"), 
    Pkg.PackageSpec(name="ImageIO", version="0.4.1"), 
])

using Images

default_iocontext = IOContext(devnull, :color => false, :limit => true, :displaysize => (18, 88))

io() = IOContext(IOBuffer(), default_iocontext)

img = load(download("https://user-images.githubusercontent.com/6933510/107239146-dcc3fd00-6a28-11eb-8c7b-41aaf6618935.png"))
small = img[1:7,1:7]

for i in 1:10000
    show(io(), MIME"image/png"(), small)
end
```

Result:
```
fons@rekenmachine Documents % julia imageieosegfault.jl
 Activating new environment at `/var/folders/v_/fhpj9jn151d4p9c2fdw2gv780000gn/T/jl_OojTpv/Project.toml`
   Updating registry at `~/.julia/registries/General`
  Resolving package versions...
Updating `/private/var/folders/v_/fhpj9jn151d4p9c2fdw2gv780000gn/T/jl_OojTpv/Project.toml`
  [82e4d734] + ImageIO v0.4.1
  [916415d5] + Images v0.23.3
Updating `/private/var/folders/v_/fhpj9jn151d4p9c2fdw2gv780000gn/T/jl_OojTpv/Manifest.toml`
  [621f4979] + AbstractFFTs v1.0.0
  [79e6a3ab] + Adapt v3.2.0
  [4fba245c] + ArrayInterface v3.1.1
  [56f22d72] + Artifacts v1.3.0
  [13072b0f] + AxisAlgorithms v1.0.0
  [39de3d68] + AxisArrays v0.4.3
  [fa961155] + CEnum v0.4.1
  [aafaddc9] + CatIndices v0.2.2
  [d360d2e6] + ChainRulesCore v0.9.29
  [3da002f7] + ColorTypes v0.10.9
  [c3611d14] + ColorVectorSpace v0.8.7
  [5ae59095] + Colors v0.12.6
  [34da2185] + Compat v3.25.0
  [e66e0078] + CompilerSupportLibraries_jll v0.3.4+0
  [ed09eef8] + ComputationalResources v0.3.2
  [150eb455] + CoordinateTransformations v0.6.1
  [dc8bdbbb] + CustomUnitRanges v1.0.1
  [9a962f9c] + DataAPI v1.5.1
  [864edb3b] + DataStructures v0.18.9
  [b4f34e82] + Distances v0.10.2
  [da5c29d0] + EllipsisNotation v1.1.0
  [4f61f5a4] + FFTViews v0.3.1
  [7a1cc6ca] + FFTW v1.3.2
  [f5851436] + FFTW_jll v3.3.9+7
  [5789e2e9] + FileIO v1.4.5
  [53c48c17] + FixedPointNumbers v0.8.4
  [a2bd30eb] + Graphics v1.1.0
  [bbac6d45] + IdentityRanges v0.3.1
  [615f187c] + IfElse v0.1.0
  [2803e5a7] + ImageAxes v0.6.8
  [f332f351] + ImageContrastAdjustment v0.3.6
  [a09fc81d] + ImageCore v0.8.20
  [51556ac3] + ImageDistances v0.2.10
  [6a3955dd] + ImageFiltering v0.6.19
  [82e4d734] + ImageIO v0.4.1
  [bc367c6b] + ImageMetadata v0.9.4
  [787d08f9] + ImageMorphology v0.2.9
  [2996bd0c] + ImageQualityIndexes v0.2.1
  [4e3cecfd] + ImageShow v0.2.3
  [02fcd773] + ImageTransformations v0.8.9
  [916415d5] + Images v0.23.3
  [9b13fd28] + IndirectArrays v0.5.1
  [1d5cc7b8] + IntelOpenMP_jll v2018.0.3+2
  [a98d9a8b] + Interpolations v0.13.1
  [8197267c] + IntervalSets v0.5.2
  [c8e1da08] + IterTools v1.3.0
  [692b3bcd] + JLLWrappers v1.2.0
  [4af54fe1] + LazyArtifacts v1.3.0
  [856f044c] + MKL_jll v2021.1.1+1
  [1914dd2f] + MacroTools v0.5.6
  [dbb5928d] + MappedArrays v0.3.0
  [e1d29d7a] + Missings v0.4.5
  [e94cdb99] + MosaicViews v0.2.4
  [77ba4419] + NaNMath v0.3.5
  [f09324ee] + Netpbm v1.0.0
  [6fe1bfb0] + OffsetArrays v1.5.3
  [efe28fd5] + OpenSpecFun_jll v0.5.3+4
  [bac558e1] + OrderedCollections v1.3.3
  [f57f5aa1] + PNGFiles v0.3.5
  [5432bcbf] + PaddedViews v0.5.8
  [d96e819e] + Parameters v0.12.2
  [b3c3ace0] + RangeArrays v0.3.2
  [c84ed2f1] + Ratios v0.4.0
  [189a3867] + Reexport v1.0.0
  [ae029012] + Requires v1.1.2
  [6038ab10] + Rotations v1.0.2
  [699a6c99] + SimpleTraits v0.9.3
  [a2af1166] + SortingAlgorithms v0.3.1
  [276daf66] + SpecialFunctions v1.2.1
  [90137ffa] + StaticArrays v1.0.1
  [2913bbd2] + StatsBase v0.33.2
  [06e1c1a7] + TiledIteration v0.2.5
  [3a884ed6] + UnPack v1.0.2
  [efce3f68] + WoodburyMatrices v0.5.3
  [83775a58] + Zlib_jll v1.2.11+18
  [b53b4c65] + libpng_jll v1.6.37+6
  [2a0f44e3] + Base64
  [ade2ca70] + Dates
  [8bb1440f] + DelimitedFiles
  [8ba89e20] + Distributed
  [b77e0a4c] + InteractiveUtils
  [76f85450] + LibGit2
  [8f399da3] + Libdl
  [37e2e46d] + LinearAlgebra
  [56ddb016] + Logging
  [d6f4376e] + Markdown
  [a63ad114] + Mmap
  [44cfe95a] + Pkg
  [de0858da] + Printf
  [3fa0cd96] + REPL
  [9a3f8284] + Random
  [ea8e919c] + SHA
  [9e88b42a] + Serialization
  [1a1011a3] + SharedArrays
  [6462fe0b] + Sockets
  [2f01184e] + SparseArrays
  [10745b16] + Statistics
  [8dfed614] + Test
  [cf7118a7] + UUIDs
  [4ec0a83e] + Unicode


signal (11): Segmentation fault: 11
in expression starting at /Users/fons/Documents/imageieosegfault.jl:19
jl_object_id_ at /Users/julia/buildbot/worker/package_macos64/build/src/builtins.c:349
typekey_hash at /Users/julia/buildbot/worker/package_macos64/build/src/jltypes.c:1014 [inlined]
jl_precompute_memoized_dt at /Users/julia/buildbot/worker/package_macos64/build/src/jltypes.c:1048
inst_datatype_inner at /Users/julia/buildbot/worker/package_macos64/build/src/jltypes.c:1314
jl_apply_tuple_type_v_ at /Users/julia/buildbot/worker/package_macos64/build/src/jltypes.c:1385 [inlined]
jl_apply_tuple_type_v at /Users/julia/buildbot/worker/package_macos64/build/src/jltypes.c:1395
arg_type_tuple at /Users/julia/buildbot/worker/package_macos64/build/src/gf.c:1851
jl_lookup_generic_ at /Users/julia/buildbot/worker/package_macos64/build/src/gf.c:2361
jl_apply_generic at /Users/julia/buildbot/worker/package_macos64/build/src/gf.c:2394
_writecallback at /Users/fons/.julia/packages/PNGFiles/CQUsD/src/io.jl:422
unknown function (ip: 0x138973dbf)
png_write_data at /Users/fons/.julia/artifacts/fb2695d9f2ba598ca2213b0768aac906ebcd1000/lib/libpng16.16.37.0.dylib (unknown line)
png_write_chunk_header at /Users/fons/.julia/artifacts/fb2695d9f2ba598ca2213b0768aac906ebcd1000/lib/libpng16.16.37.0.dylib (unknown line)
png_write_complete_chunk at /Users/fons/.julia/artifacts/fb2695d9f2ba598ca2213b0768aac906ebcd1000/lib/libpng16.16.37.0.dylib (unknown line)
png_compress_IDAT at /Users/fons/.julia/artifacts/fb2695d9f2ba598ca2213b0768aac906ebcd1000/lib/libpng16.16.37.0.dylib (unknown line)
png_write_finish_row at /Users/fons/.julia/artifacts/fb2695d9f2ba598ca2213b0768aac906ebcd1000/lib/libpng16.16.37.0.dylib (unknown line)
png_write_filtered_row at /Users/fons/.julia/artifacts/fb2695d9f2ba598ca2213b0768aac906ebcd1000/lib/libpng16.16.37.0.dylib (unknown line)
png_write_find_filter at /Users/fons/.julia/artifacts/fb2695d9f2ba598ca2213b0768aac906ebcd1000/lib/libpng16.16.37.0.dylib (unknown line)
png_write_row at /Users/fons/.julia/artifacts/fb2695d9f2ba598ca2213b0768aac906ebcd1000/lib/libpng16.16.37.0.dylib (unknown line)
png_write_image at /Users/fons/.julia/artifacts/fb2695d9f2ba598ca2213b0768aac906ebcd1000/lib/libpng16.16.37.0.dylib (unknown line)
_write_image at /Users/fons/.julia/packages/PNGFiles/CQUsD/src/io.jl:426 [inlined]
#_save#15 at /Users/fons/.julia/packages/PNGFiles/CQUsD/src/io.jl:414
_save##kw at /Users/fons/.julia/packages/PNGFiles/CQUsD/src/io.jl:344
JuliaIO/ImageIO.jl#13 at /Users/fons/.julia/packages/PNGFiles/CQUsD/src/io.jl:326
maybe_lock at /Users/fons/.julia/packages/PNGFiles/CQUsD/src/io.jl:36 [inlined]
maybe_lock at /Users/fons/.julia/packages/PNGFiles/CQUsD/src/io.jl:38 [inlined]
#save#12 at /Users/fons/.julia/packages/PNGFiles/CQUsD/src/io.jl:323
save at /Users/fons/.julia/packages/PNGFiles/CQUsD/src/io.jl:315
jl_apply at /Users/julia/buildbot/worker/package_macos64/build/src/./julia.h:1690 [inlined]
do_apply at /Users/julia/buildbot/worker/package_macos64/build/src/builtins.c:655
jl_f__apply at /Users/julia/buildbot/worker/package_macos64/build/src/builtins.c:669 [inlined]
jl_f__apply_latest at /Users/julia/buildbot/worker/package_macos64/build/src/builtins.c:705
#invokelatest#1 at ./essentials.jl:710
jl_apply at /Users/julia/buildbot/worker/package_macos64/build/src/./julia.h:1690 [inlined]
do_apply at /Users/julia/buildbot/worker/package_macos64/build/src/builtins.c:655
invokelatest at ./essentials.jl:709
jl_apply at /Users/julia/buildbot/worker/package_macos64/build/src/./julia.h:1690 [inlined]
do_apply at /Users/julia/buildbot/worker/package_macos64/build/src/builtins.c:655
#save#7 at /Users/fons/.julia/packages/ImageIO/uPn5K/src/ImageIO.jl:46
save##kw at /Users/fons/.julia/packages/ImageIO/uPn5K/src/ImageIO.jl:41
inner at ./essentials.jl:713
unknown function (ip: 0x138984e2c)
jl_apply at /Users/julia/buildbot/worker/package_macos64/build/src/./julia.h:1690 [inlined]
do_apply at /Users/julia/buildbot/worker/package_macos64/build/src/builtins.c:655
jl_f__apply at /Users/julia/buildbot/worker/package_macos64/build/src/builtins.c:669 [inlined]
jl_f__apply_latest at /Users/julia/buildbot/worker/package_macos64/build/src/builtins.c:705
#invokelatest#1 at ./essentials.jl:714 [inlined]
invokelatest##kw at ./essentials.jl:709 [inlined]
#save#33 at /Users/fons/.julia/packages/FileIO/TyKdX/src/loadsave.jl:233
unknown function (ip: 0x13898349a)
save##kw at /Users/fons/.julia/packages/FileIO/TyKdX/src/loadsave.jl:217
unknown function (ip: 0x138982667)
#show#13 at /Users/fons/.julia/packages/ImageShow/9kpaq/src/showmime.jl:43
unknown function (ip: 0x138980478)
show at /Users/fons/.julia/packages/ImageShow/9kpaq/src/showmime.jl:28
unknown function (ip: 0x13897a456)
top-level scope at /Users/fons/Documents/imageieosegfault.jl:20
jl_toplevel_eval_flex at /Users/julia/buildbot/worker/package_macos64/build/src/toplevel.c:834
jl_parse_eval_all at /Users/julia/buildbot/worker/package_macos64/build/src/ast.c:913
jl_load_rewrite at /Users/julia/buildbot/worker/package_macos64/build/src/toplevel.c:914 [inlined]
jl_load at /Users/julia/buildbot/worker/package_macos64/build/src/toplevel.c:919
include at ./Base.jl:380
include at ./Base.jl:368
exec_options at ./client.jl:296
_start at ./client.jl:506
jfptr__start_57081 at /Applications/Julia-1.5.app/Contents/Resources/julia/lib/julia/sys.dylib (unknown line)
true_main at /usr/local/bin/julia (unknown line)
main at /usr/local/bin/julia (unknown line)
Allocations: 43572662 (Pool: 43555600; Big: 17062); GC: 44
zsh: segmentation fault  julia imageieosegfault.jl
```

System info:

```
julia> versioninfo()
Julia Version 1.5.3
Commit 788b2c77c1 (2020-11-09 13:37 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.7.0)
  CPU: Apple M1
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, westmere)
```

**Comment by IanButterworth** [comment]
I brought this over from ImageIO

Reduced one small step further. The IOContext kwargs aren't needed. This segfaults

```julia
using PNGFiles
begin
  img = PNGFiles.load(download("https://user-images.githubusercontent.com/6933510/107239146-dcc3fd00-6a28-11eb-8c7b-41aaf6618935.png"))
  small = img[1:7,1:7]
  default_iocontext = IOContext(devnull)
  for i in 1:500_000
      PNGFiles.save(IOContext(IOBuffer(), default_iocontext), small)
  end
end

**Comment by IanButterworth** [comment]
As a `Base` sanity check, this doesn't segfault
```julia
begin
    img = rand(UInt8, 7, 7)
    for i in 1:1_000_000
         write(IOBuffer(), img)
    end
    println("first loop done")
    for i in 1:1_000_000
         write(IOContext(IOBuffer()), img)
    end
end
```

**Comment by IanButterworth** [comment]
Potential work-around/fix here https://github.com/JuliaIO/PNGFiles.jl/pull/33

On that PR, both loops complete.

**Comment by IanButterworth** [comment]
Ok, this should be fixed in `v0.3.6` that was just released. I'll close the issue, but please say if it's not fixed. or perhaps even if it is :)
